google/cloud/internal/tracing_http_payload.cc

```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/internal/tracing_http_payload.h"
+#include "google/cloud/internal/opentelemetry.h"
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+TracingHttpPayload::TracingHttpPayload(
+    std::unique_ptr<HttpPayload> impl,
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span)
+    : impl_(std::move(impl)), span_(std::move(span)) {}
+
+bool TracingHttpPayload::HasUnreadData() const {
+  return impl_->HasUnreadData();
+}
+
+StatusOr<std::size_t> TracingHttpPayload::Read(absl::Span<char> buffer) {
+  auto status = impl_->Read(buffer);
+  if (!status) {
+    span_->AddEvent(
+        "Read", {{"read.buffer.size", static_cast<std::int64_t>(buffer.size())},
+                 {"read.error", true}});
+    return internal::EndSpan(*span_, std::move(status));
+  }
+  span_->AddEvent(
+      "Read", {{"read.buffer.size", static_cast<std::int64_t>(buffer.size())},
+               {"read.returned.size", static_cast<std::int64_t>(*status)}});
+  if (*status == 0) span_->End();
```

Comment:
I would recommend:

```cc
if (*status == 0) return EndSpan(span_, std::move(status));
```

I'd rather we hit this code path in all of the code, even if we decide that we don't want to add attributes in the success case.

https://github.com/googleapis/google-cloud-cpp/blob/fba937ed589c3a118f712722d93e32c1486c235e/google/cloud/internal/opentelemetry.cc#L52-L57

Aside: You might have read that OpenTelemetry suggests leaving the Span status as `kUnset` in the success case. I have been setting them to `kOk`. I am not worried about removing this line of code. I will just need to update a bunch of tests. https://github.com/googleapis/google-cloud-cpp/blob/fba937ed589c3a118f712722d93e32c1486c235e/google/cloud/internal/opentelemetry.cc#L53

Reply:
Done